session 5 persistence

misamoto opened the pull request:
feat: setup h2 in-memory db
 - include dependencies for h2, jdbc (and devtools)
 - define datasource props
 - setup database initialization on app start-up
 - create schema generation script
 - create data generation script
 - set 'spring.jpa.hibernate.ddl-auto' property to none
   to stop it getting into way of db scripts generation

feat: implement db repositories
 - create common abstract db repository class to provide common functionality
 - add transactions on service layer

rapasoft commented on src/main/java/no/itera/bloggingplatform/repository/database/DbAuthorRepository.java:
```diff
@@ -0,0 +1,99 @@
+
+  @Override
+  public Author update(Author toUpdate) {
+    delete(toUpdate.getKey());
```
This will change the keys - it is not consistent with the inMemory repo (where the key is the same).

misamoto replied:
So true ... got tied up trying to have a composite operation, where I'd show transactions.
Will fix this.

rapasoft commented on src/main/java/no/itera/bloggingplatform/repository/database/DbPostRepository.java:
```diff
@@ -0,0 +1,142 @@
+        toPersist.getDate(),
+        toPersist.getPerex(),
+        toPersist.getContent(),
+        toPersist.getAuthor().getKey()
```
InMemory DB perists author as well (here you expect ID is already set)

misamoto replied:
`Author` is persisted and set to `post` on service layer, prior to `postRepository.create(post)` call.

rapasoft replied:
Ah, sure, ok.

rapasoft commented on src/main/java/no/itera/bloggingplatform/repository/database/DbPostRepository.java:
```diff
@@ -0,0 +1,142 @@
+
+  private static Post mapToPost(
+      ResultSet resultSet,
+      AuthorRepository authorRepository,
```
I don't think putting repository as parameter for mapper is a good idea. We should rather compose mappers

misamoto replied:
`compose mappers` 
.. not sure what do you mean here ..

Commit:
feat: setup h2 in-memory db

- include dependencies for h2, jdbc (and devtools)
- define datasource props
- setup database initialization on app start-up
- create schema generation script
- create data generation script

Commit:
feat: implement db repositories

- create common abstract db repository class to provide common functionality
- add transactions on service layer
